eyetracking/bids_scripts/et_prep_step1.py

```diff
@@ -0,0 +1,429 @@
+    '''
+    sub, ses, run, task, fnum = pupil_path[1]
+
+    task_root = out_path.split('/')[-1]
+    pseudo_task = 'task-mario3' if task_root == 'mario3' else task
+
+    outpath_gaze = os.path.join(out_path, sub, ses)
+    gfile_path = f'{outpath_gaze}/{sub}_{ses}_{run}_{fnum}_{pseudo_task}_gaze2D.npz'
+
+    if not os.path.exists(gfile_path):
+        # gaze data includes pupil metrics from which each gaze was derived
+        seri_gaze = load_pldata_file(pupil_path[0], 'gaze')[0]
+        print(sub, ses, run, pseudo_task, len(seri_gaze))
+
+        # Convert serialized file to list of dictionaries
+        gaze_2plot_list = []
+        deserialized_gaze = []
+
+        for gaze in seri_gaze:
+            gaze_data = {}
+            # [gaze_x, gaze_y, pupil_x, pupil_y, timestamp, confidence]
+            gaze_2plot = np.empty(6)
+            for key in gaze.keys():
+                if key != 'base_data': # gaze data
+                    if key == 'norm_pos':
+                        gaze_2plot[0: 2] = [gaze[key][0], gaze[key][1]]
+                    elif key == 'timestamp':
+                        gaze_2plot[4] = gaze[key]
+                    elif key == 'confidence':
+                        gaze_2plot[5] = gaze[key]
+                    gaze_data[key] = gaze[key]
+                else: # pupil data from which gaze was derived
+                    gaze_pupil_data = {}
+                    gaze_pupil = gaze[key][0]
+                    for k in gaze_pupil.keys():
+                        if k != 'ellipse':
+                            if k == 'norm_pos':
+                                gaze_2plot[2: 4] = [
+                                    gaze_pupil[k][0],
+                                    gaze_pupil[k][1],
+                                ]
+                            gaze_pupil_data[k] = gaze_pupil[k]
+                        else:
+                            gaze_pupil_ellipse_data = {}
+                            for sk in gaze_pupil[k].keys():
+                                gaze_pupil_ellipse_data[sk] = gaze_pupil[k][sk]
+                            gaze_pupil_data[k] = gaze_pupil_ellipse_data
+                    gaze_data[key] = gaze_pupil_data
+
+            deserialized_gaze.append(gaze_data)
+            gaze_2plot_list.append(gaze_2plot)
+
+        print(len(deserialized_gaze))
+
+        try:
+            if task_root == 'friends':
+                run_dur = 770
+            else:
+                if task_root in ['retino', 'floc']:
+                    ev_path = f'{in_path}/{sub}/{ses}/{sub}_{ses}_{fnum}_{task}_events.tsv'
+                else:
+                    ev_path = f'{in_path}/{sub}/{ses}/{sub}_{ses}_{fnum}_{task}_{run}_events.tsv'
+                ev_lasttrial = pd.read_csv(ev_path, sep='\t', header=0).iloc[-1]
+                run_dur = int(ev_lasttrial['onset'] + 20)
+        except:
+            print('event file did not load, using default run duration')
+            run_dur = 700
```

Comment:
what?!

Reply:
yeah... that's just for plotting, to set the length of the x axis to include the full run, based on the last trial's onset on the event.tsv file. Friends doesn't have events.tsv files, and then a few runs from the other tasks also had that file missing, so I used 700s (11min67s, to include a Friends run) in those cases.